modules/cloudflarepurger/src/Plugin/Purge/Purger/CloudFlarePurger.php

```diff
@@ -0,0 +1,210 @@
+
+    parent::__construct($configuration, $plugin_id, $plugin_definition);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function invalidate(InvalidationInterface $invalidation) {
+    $this->invalidateMultiple([$invalidation]);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function invalidateMultiple(array $invalidations) {
+    // @todo figure out why this is happening.
+    if (count($invalidations) == 0) {
+      $this->logger->warning("No Invalidations?  That's odd." . print_r($invalidations, TRUE));
+      return;
+    }
+
+    $chunks = array_chunk($invalidations, CloudFlareAPI::MAX_TAG_PURGES_PER_REQUEST);
+    foreach ($chunks as $chunk) {
+      $this->purgeChunk($chunk);
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getTimeHint() {
+    return 1.3;
+  }
+
+  /**
+   * Purges a chunk of tags.
+   *
+   * Integration point between purge and CloudFlareAPI.  Purge requires state
+   * tracking on each item purged.  This function provides that accounting and
+   * calls CloudflareApi.
+   *
+   * CloudFlare only allows us to purge 30 tags at once.
+   *
+   * @param array $invalidations
+   *   Chunk of purge module invalidation objects to purge via CloudFlare.
+   */
+  private function purgeChunk(array &$invalidations) {
+    $api_targets_to_purge = [];
+
+    // This method is unfortunately a bit verbose due to the fact that we
+    // need to update the purge states as we proceed.
+    foreach ($invalidations as $invalidation) {
+      $invalidation->setState(InvalidationInterface::STATE_PURGING);
+      $api_targets_to_purge[] = $invalidation->getExpression();
+    }
+
+    try {
+      // Interface with the CloudFlarePhpSdk.
+      $invalidation_type = $invalidations[0]->getPluginId();
+      if ($invalidation_type == 'tag') {
+        // @todo Remove this wrapper once CloudFlare supports 16k headers.
+        // Also invalidate the cache tags as hashes, to automatically also work
+        // for responses that exceed CloudFlare's Cache-Tag header limit.
+        $hashes = CloudFlareCacheTagHeaderGenerator::cacheTagsToHashes($api_targets_to_purge);
+        $tags = Cache::mergeTags($api_targets_to_purge, $hashes);
+
+        $this->zoneApi->purgeTags($this->zone, $tags);
+        $this->state->incrementTagPurgeDailyCount();
+      }
+
+      elseif ($invalidation_type == 'url') {
+        $this->zoneApi->purgeIndividualFiles($this->zone, $api_targets_to_purge);
+      }
+
+      elseif ($invalidation_type == 'everything') {
+        $this->zoneApi->purgeAllFiles($this->zone);
+      }
+
+      foreach ($invalidations as $invalidation) {
+        $invalidation->setState(InvalidationInterface::STATE_PURGED);
+      }
+    }
+
+    catch (\Exception $e) {
+      foreach ($invalidations as $invalidation) {
+        $this->logger->error($e->getMessage());
+        $invalidation->setState(InvalidationInterface::STATE_FAILED);
+      }
+    }
+
+    finally {
```

Comment:
Nifty, I never heard about this statement... it feels semi-pointless though: *...within the finally block will always be executed after the try and catch blocks, regardless of whether an exception has been thrown, and before normal execution resumes.*

So that replaces a newline? ;)